src/Shared/TaskLoggingHelper.cs

```diff
@@ -258,6 +258,33 @@ public bool LogsMessagesOfImportance(MessageImportance importance)
                 || buildEngine10.EngineServices.LogsMessagesOfImportance(importance);
         }
 
+#if NET6_0_OR_GREATER
+
+        /// <summary>
+        /// Logs a message using the specified interpolated string.
+        /// Thread safe.
+        /// </summary>
+        /// <param name="stringHandler">The message interpolated string.</param>
+        public void LogMessage(LogInterpolatedStringHandler stringHandler)
+        {
+            LogMessage(MessageImportance.Normal, stringHandler);
+        }
+
+        /// <summary>
+        /// Logs a message using the specified interpolated string.
+        /// Thread safe.
+        /// </summary>
+        /// <param name="importance">The importance level of the message.</param>
+        /// <param name="stringHandler">The message interpolated string.</param>
+        public void LogMessage(MessageImportance importance, LogInterpolatedStringHandler stringHandler)
+        {
+            if (LogsMessagesOfImportance(importance))
+            {
+                LogMessage(importance, stringHandler.GetFormat(), stringHandler.Arguments);
```

Comment:
A great thing about InterpolatedStringHandlers is that you can move these `if` conditions into the handler itself, and when the condition is false, you can opt-out of any processing of the string at all.

See https://github.com/dotnet/runtime/blob/a4c92435ebc331d4f5e107c37d2fe759a3199a32/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Debug.cs#L274-L291 for an example.

You can change this to be:

```C#
public void LogMessage(MessageImportance importance, [InterpolatedStringHandlerArgument("", nameof(importance))] LogInterpolatedStringHandler stringHandler)
{
```

and then in the constructor of `LogInterpolatedStringHandler`:

```C#
public LogInterpolatedStringHandler(int literalLength, int formattedCount, TaskLoggingHelper taskLoggingHelper, MessageImportance importance, out bool shouldAppend)
{
    if (taskLoggingHelper.LogsMessagesOfImportance(importance))
    {
        // Do normal processing/allocating
        shouldAppend = true;
    }
    else
    {
        shouldAppend = false;
    }
}
```

cc @stephentoub

Reply:
Thank you, how could we have missed this!

If we frontload the importance check to our interpolated string handler constructor, we don't have to optimize for the case where `AppendLiteral` and `AppendFormatted` have been called but the string is not needed at the end. This radically changes the design of the handler. Definitely make sense to compose our handler out of a default one and just add the early check, just like in `Debug.Write*`.

(Note: this PR is being reverted in #8479 because the requirement to run MSBuild tasks built against this change on MSBuild with this change resulted in source build break. We'll have to redo this anyway.)

Reply:
FWIW...

You're reverting it anyway, so it doesn't really matter, but as I mentioned offline to @eerhardt, I wouldn't be surprised if the new version (without the severity check) was also more expensive in many situations than if you didn't have the custom handler at all.  If someone just wrote an interpolated string and that got compiled down using the default handler, you'd get a single string allocation in the common case.  With the interpolated handler here, you're allocating one or more char[]s in order to then produce a string, you're allocating an object[] to store all the arguments, you're boxing each value type argument to store into that object[], etc.  If that's really the path you wanted to go, because you needed a composite format string and the object[] for some other reason (e.g. maybe the logging infrastructure requires the ability to perform custom conversions on the args or something), you could just use the old-school `FormattableString`, where the compiler generates the composite format string for you at compile time and handles the creation of the object[].  But, if it's fine to just produce the result string after a severity check, you're much better off doing what was discussed above, and having the handler do the severity check and then delegate to the default handler.

Reply:
Thanks Eric and Stephen. We had a plan to do more optimizations for the handler. I wanted to use ```FormattableString```, but compiler was favorizing the old overload with ```params object[]```. The solution would be use new method name (which would probably prevent the issue with source build break), but I didn't like the idea as it could be confusing for developers. Now we have opportunity to take a second look with fresh mind and gained experience :) And we should probably review our whole logging pipeline, and make more fundamental changes to support interpolated strings.